qiskit_ode/signals/signals.py

```diff
@@ -27,167 +28,196 @@
 from qiskit_ode.dispatch import Array
 
 
-class BaseSignal(ABC):
-    """Base class for a time-dependent mixed signal."""
+class Signal:
+    r"""General signal class representing a function of the form:
 
-    def __init__(self, name: str = None):
-        """Init function."""
+    .. math::
+        Re[f(t)e^{i (2 \pi \nu t + \phi)}]
+                = Re[f(t)]\cos(2 \pi \nu t + \phi) - Im[f(t)]\sin(2 \pi \nu t + \phi),
+
+    where
+
+    - :math:`f(t)` is the envelope function.
+    - :math:`\nu` is the carrier frequency.
+    - :math:`\phi` is the phase.
+
+    The envelope function can be complex-valued, and the frequency and phase must be real.
+
+    Note: this class assumes that the envelope function is vectorized. If it isn't, it can
+    be vectorized automatically by calling ``numpy.vectorize``, or, if using JAX, by
+    calling ``jax.numpy.vectorize``.
+    """
+
+    def __init__(
+        self,
+        envelope: Union[Callable, complex, float, int],
+        carrier_freq: float = 0.0,
+        phase: float = 0.0,
+        name: Optional[str] = None,
+    ):
+        """
+        Initializes a signal given by an envelope and a carrier.
+
+        Args:
+            envelope: Envelope function of the signal, must be vectorized.
+            carrier_freq: Frequency of the carrier.
+            phase: The phase of the carrier.
+            name: Name of the signal.
+        """
         self._name = name
-        self._carrier_freq = Array(0.0)
-        self._phase = Array(0.0)
 
-    @property
-    def carrier_freq(self):
-        """Return the carrier frequency of the signal."""
-        return self._carrier_freq
+        if isinstance(envelope, (float, int)):
+            envelope = Array(complex(envelope))
 
-    @property
-    def phase(self):
-        """Return the phase of the Signal."""
-        return self._phase
+        if isinstance(envelope, Array):
+            self._envelope = lambda t: envelope
 
-    @abstractmethod
-    def conjugate(self):
-        """Return a new signal that is the complex conjugate of self."""
+        if callable(envelope):
+            self._envelope = envelope
 
-    @abstractmethod
-    def envelope_value(self, t: float = 0.0) -> complex:
-        """Evaluates the envelope at time t."""
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
```

Comment:
Is this needed if you set them immediately after? Or does lint or the editor complain?

Reply:
Just removed these lines and the linter doesn't complain, so I'll get rid of them!